src/statement-list/converters/expr2statement_list.cpp

```diff
@@ -25,3 +117,187 @@ std::string type2stl(const typet &type, const namespacet &ns)
+  convert_multiary_bool(operands, XOR);
+}
+
+void expr2stlt::convert(const not_exprt &expr)
+{
+  // If a NOT expression is being handled here it must always mark the
+  // beginning of a new bit string.
+  PRECONDITION(!inside_bit_string);
+
+  if(ID_symbol == expr.op().id())
+  {
+    // Use AN to load the symbol.
+    is_reference = true;
+    result << AND << NOT_POSTFIX << OPERAND_SEPARATOR;
+    convert(to_symbol_expr(expr.op()));
+    result << LINE_SEPARATOR;
+  }
+  else
+  {
+    // Use NOT to negate the RLO after the wrapped expression was loaded.
+    convert(expr.op());
+    result << NOT LINE_SEPARATOR;
+  }
+}
+
+void expr2stlt::convert(const symbol_exprt &expr)
+{
+  if(is_reference)
+  {
+    result << REFERENCE_FLAG;
+    is_reference = false;
+  }
+  result << id2string(id_shorthand(expr.get_identifier()));
+}
+
+void expr2stlt::convert_multiary_bool(
+  std::vector<exprt> &operands,
+  const char operation)
+{
+  if(inside_bit_string)
+    convert_multiary_bool_operands(operands, operation);
+  else
+  {
+    convert_first_non_trivial_operand(operands);
+    convert_multiary_bool_operands(operands, operation);
+  }
+}
+
+void expr2stlt::convert_multiary_bool_operands(
+  const std::vector<exprt> &operands,
+  const char operation)
+{
+  for(const exprt &op : operands)
+  {
+    if(ID_not == op.id())
+    {
+      result << operation << NOT_POSTFIX;
+      convert_bool_operand(op.op0());
+    }
+    else
+    {
+      result << operation;
+      convert_bool_operand(op);
+    }
+  }
+}
+
+void expr2stlt::convert_bool_operand(const exprt &op)
+{
+  if(op.id() == ID_symbol)
+  {
+    is_reference = true;
+    result << OPERAND_SEPARATOR;
+    convert(op);
+    result << LINE_SEPARATOR;
+  }
+  else
+  {
+    inside_bit_string = false;
+    result << NESTING_OPEN_LINE_SEPARATOR;
+    convert(op);
+    result << NESTING_CLOSED_LINE_SEPARATOR;
+  }
+}
+
+void expr2stlt::convert_first_non_trivial_operand(std::vector<exprt> &operands)
+{
+  exprt non_trivial_op;
+  for(auto it(begin(operands)); it != end(operands); ++it)
+  {
+    if(
+      (ID_symbol == it->id()) ||
+      (ID_not == it->id() && ID_symbol == it->op0().id()))
+      continue;
+    else
+    {
+      non_trivial_op = *it;
+      operands.erase(it);
+      break;
+    }
+  }
+  // Important for some scenarios: Convert complex op first, set bit string
+  // flag to true afterwards.
+  if(non_trivial_op.is_not_nil())
+    convert(non_trivial_op);
+
+  inside_bit_string = true;
+}
+
+irep_idt expr2stlt::id_shorthand(const irep_idt &identifier)
+{
+  const symbolt *symbol;
+  std::string shorthand = id2string(identifier);
+  if(
+    !ns.lookup(identifier, symbol) && !symbol->base_name.empty() &&
+    has_suffix(shorthand, id2string(symbol->base_name)))
+    return symbol->base_name;
+
+  const std::string::size_type pos = shorthand.rfind(SCOPE_SEPARATOR);
+  if(pos != std::string::npos)
+    shorthand.erase(0, pos + 2);
```

Comment:
Prefer `strlen(SCOPE_SEPARATOR)`. It is self-documenting and since `SCOPE_SEPARATOR` is a `#define` constant, the compiler replaces it by the constant `2`.